### PR TITLE
Change Outbox to use synthetic partition key

### PR DESCRIPTION
**Pull request by jpalac**
Fixes
- https://github.com/Particular/NServiceBus.Persistence.CosmosDB/issues/893
- https://github.com/Particular/NServiceBus.Persistence.CosmosDB/issues/899

### CODE REVIEW

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 1]
```diff
@@ -29,26 +40,85 @@ public async Task<OutboxMessage> Get(string messageId, ContextBag context, Cance
...
 
-        OutboxRecord outboxRecord = await setAsDispatchedHolder.ContainerHolder.Container.ReadOutboxRecord(messageId, partitionKey, serializer, cancellationToken)
+        // If the user has overridden the default synthetic partition key strategy, then partitionKeyObject will be what they have set. If not, it will be the synthetic.
+        OutboxRecord outboxRecord = await setAsDispatchedHolder.ContainerHolder.Container.ReadOutboxRecord(messageId, finalPartitionKey, serializer, cancellationToken)
             .ConfigureAwait(false);
 
+        // Only attempt the fallback if the user has NOT overridden the partition key strategy and the readFallbackEnabled flag is set.
+        // Theres no point in trying to fallback if the user has specified their own partition key strategy and the record wasn't found.
+        // This saves an unnecessary read.
+        if (outboxRecord is null && readFallbackEnabled && !extractorConfig.HasAnyCustomPartitionExtractors)
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:102]: Example https://github.com/Particular/NServiceBus.Persistence.CosmosDB/blob/master/src/NServiceBus.Persistence.CosmosDB/Transaction/TransactionInformationBeforeTheLogicalOutboxBehavior.cs#L20 right above this line you can set a result key to true and that then can then be observed at any lower stage. Same applies for header extraction and container at various stages. So essentially something like PartitionKeyExtracted = true would suffice

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 2]
```diff
@@ -1,54 +1,129 @@
...
 
+        // Only set partition key if:
+        // 1. We have one in context AND
+        // 2. We're not going to defer to logical stage for container extraction
         if (context.TryGet(out PartitionKey partitionKey))
         {
-            cosmosOutboxTransaction.PartitionKey = partitionKey;
+            // Check if we'll defer for container extraction
+            var containerHolder = containerHolderResolver.ResolveAndSetIfAvailable(context);
+            bool willDeferForContainer = containerHolder == null && extractorConfig.HasCustomContainerMessageExtractors;
```
Comment (jpalac) [inline src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:28]: Does it matter if the container is set here? ie if there is a custom extractor shouldn't we defer so that it's used?

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 3]
```diff
@@ -1,54 +1,129 @@
...
+        else
+        {
+            // We have a partition key from context (headers or previous extraction)
+            finalPartitionKey = extractedPartitionKey;
+        }
 
-            partitionKey = new PartitionKey(messageId);
-            context.Set(partitionKey);
+        // Check if we need to defer for container extraction
+        if (!setAsDispatchedHolder.ContainerIsSet() && extractorConfig.HasCustomContainerMessageExtractors)
```
Comment (jpalac) [inline src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:79]: Same comment as above - should the extractors not override defaults?

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 4]
```diff
@@ -1,54 +1,129 @@
...
+        {
+            // When deferring for container extraction, ensure partition key is set in context
+            // Use the default synthetic key if no custom partition extractors are configured
+            if (!havePartitionKeyInContext && !extractorConfig.HasAnyCustomPartitionExtractors)
+            {
+                finalPartitionKey = GetPartitionKey(extractedPartitionKey, messageId);
```
Comment (jpalac) [inline src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:85]: I don't think this is needed as the finalPartitionKey is already set above to what it should be?

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 5]
```diff
@@ -1,54 +1,129 @@
...
+                context.Set(finalPartitionKey);
+            }
+            return null;
```
Comment (jpalac) [inline src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:88]: I would set `shouldDeferToLogicalStage `to true here so that we're only returning `null `in one spot below

File: src/SharedAcceptanceTests.RequirePartitionKey/When_subscribers_handles_the_same_event.cs [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+
+public class When_subscribers_handles_the_same_event : NServiceBusAcceptanceTest
+{
+    [Test]
+    public async Task Should_be_processed_by_all_subscribers_using_default_synthetic_key()
+    {
+        Requires.OutboxPersistence();
+
+        var runSettings = new RunSettings();
+        runSettings.DoNotRegisterDefaultPartitionKeyProvider();
```
Comment (jpalac) [inline src/SharedAcceptanceTests.RequirePartitionKey/When_subscribers_handles_the_same_event.cs:21]: We changed this but I've seen that this test is [included in Core](https://github.com/Particular/NServiceBus/blob/4cce8c188f4efd5eea3358ca8adc740cf9a9d4b7/src/NServiceBus.AcceptanceTests/Outbox/When_subscribers_handles_the_same_event.cs#L15), with the intention of it being run from there as opposed to having it included in each persistence. So may need to look into this further.

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs (hunk 5 above)
Comment (warwickschroeder) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:88]: Done

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs (hunk 3 above)
Comment (warwickschroeder) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:79]: I've updated this section with additional logic when there is a message and header container extractor. With this addition we'll need the `!setAsDispatchedHolder.ContainerIsSet()` check.

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 7]
```diff
@@ -1,54 +1,132 @@
...
+            shouldDeferToLogicalStage = true;
         }
 
+        // If we determined we should defer for partition key extraction, do so now
+        if (shouldDeferToLogicalStage)
+        {
+            return null;
+        }
+
+        // TODO: throw if the container doesnt exist in cosmosDB. This can happen if the user has a custom extractor that extracts a container that doesnt exist.
```
[the remark this answers is not in the document]
Comment (jpalac) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:100]: I think that can be removed.
```suggestion
```

File: src/NServiceBus.Persistence.CosmosDB/Transaction/Transaction.cs [hunk 8]
```diff
@@ -10,6 +12,20 @@ protected override void Setup(FeatureConfigurationContext context)
     {
         TransactionInformationConfiguration configuration = context.Settings.Get<TransactionInformationConfiguration>();
 
+        context.Services.AddSingleton(provider =>
+        {
+            return new ExtractorConfigurationHolder
+            {
+                Configuration = new ExtractorConfiguration
+                {
+                    HasCustomPartitionHeaderExtractors = configuration.PartitionKeyExtractor.HasCustomHeaderExtractors || provider.GetServices<IPartitionKeyFromHeadersExtractor>().Any(),
```
[the remark this answers is not in the document]
Comment (danielmarbach) [reply src/NServiceBus.Persistence.CosmosDB/Transaction/Transaction.cs:21]: See https://github.com/Particular/NServiceBus.Persistence.CosmosDB/pull/900. I thought showing some code might be better.

File: src/NServiceBus.Persistence.CosmosDB/Outbox/ExtractorConfigurationHolder.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+namespace NServiceBus.Persistence.CosmosDB;
+
+/// <summary>
+/// Holds the ExtractorConfiguration that is populated by the pipeline behaviors
+/// and accessed by the OutboxPersister and LogicalOutboxBehavior.
+/// </summary>
+class ExtractorConfigurationHolder
```
[the remark this answers is not in the document]
Comment (warwickschroeder) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/ExtractorConfigurationHolder.cs:7]: I agree it doesn't need to be kept as a separate holder, not now anyway. Adding this config to `TransactionInformationConfiguration`, like you've done in https://github.com/Particular/NServiceBus.Persistence.CosmosDB/pull/900, makes sense.

File: src/NServiceBus.Persistence.CosmosDB/Transaction/Transaction.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (warwickschroeder) [reply src/NServiceBus.Persistence.CosmosDB/Transaction/Transaction.cs:21]: The code in https://github.com/Particular/NServiceBus.Persistence.CosmosDB/pull/900 is cleaner, so I suggest we merge it, but perhaps we should create additional tests for the DI edge cases you are describing. I do see the situation around scoped dependencies that would end up throwing here.

File: src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs [hunk 10]
```diff
@@ -41,7 +41,7 @@ public async Task<OutboxMessage> Get(string messageId, ContextBag context, Cance
 
         var havePartitionKeyInContext = context.TryGet(out PartitionKey extractedPartitionKey);
 
-        PartitionKey finalPartitionKey = PartitionKey.Null;
+        var finalPartitionKey = PartitionKey.Null;
         bool shouldDeferToLogicalStage = false;
```
Comment (PhilBastian) [inline src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:45]:
```suggestion
        var shouldDeferToLogicalStage = false;
```
this one too 🤣

Comment (warwickschroeder) [reply src/NServiceBus.Persistence.CosmosDB/Outbox/OutboxPersister.cs:45]: Var all the things

**Review by PhilBastian — APPROVED** [review]
